packages/grid/x-data-grid-pro/src/components/headerFiltering/GridHeaderFilterItem.tsx

```diff
@@ -0,0 +1,279 @@
+}
+
+type OwnerState = DataGridProProcessedProps & GridHeaderFilterItemProps;
+
+const useUtilityClasses = (ownerState: OwnerState) => {
+  const { colDef, classes, showColumnVerticalBorder } = ownerState;
+
+  const slots = {
+    root: [
+      'columnHeader',
+      colDef.headerAlign === 'left' && 'columnHeader--alignLeft',
+      colDef.headerAlign === 'center' && 'columnHeader--alignCenter',
+      colDef.headerAlign === 'right' && 'columnHeader--alignRight',
+      'withBorderColor',
+      showColumnVerticalBorder && 'columnHeader--withRightBorder',
+    ],
+  };
+
+  return composeClasses(slots, getDataGridUtilityClass, classes);
+};
+
+const GridHeaderFilterItem = React.forwardRef<HTMLDivElement, GridHeaderFilterItemProps>(
+  (props, ref) => {
+    const {
+      colIndex,
+      height,
+      hasFocus,
+      headerFilterComponent,
+      filterOperators,
+      width,
+      headerClassName,
+      colDef,
+      headerFilterMenuRef,
+      ...other
+    } = props;
+
+    const apiRef = useGridPrivateApiContext();
+    const columnFields = gridVisibleColumnFieldsSelector(apiRef);
+    const rootProps = useGridRootProps();
+    const cellRef = React.useRef<HTMLDivElement>(null);
+    const handleRef = useForkRef(ref, cellRef);
+    const inputRef = React.useRef<HTMLInputElement>(null);
+    const buttonRef = React.useRef<HTMLButtonElement>(null);
+
+    const isEditing = unstable_gridHeaderFilteringEditFieldSelector(apiRef) === colDef.field;
+    const isMenuOpen = unstable_gridHeaderFilteringMenuSelector(apiRef) === colDef.field;
+    const filterModel = useGridSelector(apiRef, gridFilterModelSelector);
+    const item = React.useMemo(
+      () =>
+        filterModel?.items.find((i) => i.field === colDef.field && i.operator !== 'isAnyOf') ??
+        getGridFilter(colDef as any),
+      [filterModel?.items, colDef],
+    );
+
+    const currentOperator = filterOperators![0];
+
+    const InputComponent =
+      (colDef.type && TYPES_WITH_NO_FILTER_CELL.includes(colDef.type)) || !colDef.filterable
+        ? null
+        : currentOperator!.InputComponent;
+
+    const applyFilterChanges = React.useCallback(
+      (updatedItem: GridFilterItem) => {
+        apiRef.current.upsertFilterItem(updatedItem);
+      },
+      [apiRef],
+    );
+
+    React.useLayoutEffect(() => {
+      if (hasFocus && !isMenuOpen) {
+        let focusableElement = cellRef.current!.querySelector<HTMLElement>('[tabindex="0"]');
+        if (isEditing && InputComponent) {
+          focusableElement = inputRef.current;
+        }
+        const elementToFocus = focusableElement || cellRef.current;
+        elementToFocus?.focus();
+      }
+    }, [InputComponent, apiRef, hasFocus, isEditing, isMenuOpen]);
+
+    const onKeyDown = React.useCallback(
+      (event: React.KeyboardEvent) => {
+        if (isMenuOpen || isNavigationKey(event.key)) {
+          return;
+        }
+        if (isEditing) {
+          switch (event.key) {
```

Comment:
Can we rework it to have a single `switch` and do the `isEditing` checks inside each `case`? I think it will make it easier to read. It's just a suggestion.

Reply:
Thanks for the suggestion, I tried to do the same earlier, but it seemed less manageable to me as compared to this, anyhow, I'll review it again and see if we can compactly reduce one level of indentation.